### PR TITLE
Switch GH to use upwind penalty boundary correction

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Have GH use the upwind penalty boundary correction that doesn't mix quantities from both sides and will work with moving meshes.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsdeppe** [comment]
@geoffrey4444 @prayush could you please let me know if you'd be willing to review this? It implements the upwind penalty boundary correction that doesn't mix quantities between different sides of the element and works with a moving mesh.

**Comment by prayush** [comment]
> @geoffrey4444 @prayush could you please let me know if you'd be willing to review this? It implements the upwind penalty boundary correction that doesn't mix quantities between different sides of the element and works with a moving mesh.

Hi @nilsdeppe I can look at this

File: src/Evolution/Systems/CurvedScalarWave/Equations.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/CurvedScalarWave/Equations.hpp]: Typo? `W compute` -> `We compute`?

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:41]: Why beta for index here but alpha above?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Started reviewing...a couple of documentation questions. Still need to look at UpwindPenaltyCorrection.?pp, Test_UpwindPenaltyCorrection.cpp, and UpwindPenaltyCorrection.py

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:49]: I suggest clarifying that hatted indexes index the characteristic variables, while unhatted indexes index the evolved variables.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp]: "since that boundary term" -> "since the boundary term given by that equation"

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp]: GH -> generalized harmonic (GH is not an acronym defined in the help text so far)

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp]: only use a comma if the equation does not end the sentence. Not sure if you will end up with this ending the sentence, as it currently does, or will add more before ending the sentence.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:94]: Just a note: I checked by hand that these equations follow from the original definition after plugging in the char fields and speeds.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:109]: I'm curious: it looks like the mesh velocity term is just a correction to the shift, `\beta^i \rightarrow \beta^i + v^i_g`, except for this line, where you only apply the correction to the term without the constraint damping.

I'm just double-checking whether this is correct, and if so, if there is an easy way to understand why there's no term `-\gamma_1 v^i_g` here.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:126]: Clarify that here we assume that the normals point in the same direction, but in the code, they point in opposite directions. It would be even better to add to the help text the equations with the spectre, sign-flipping normals explicitly, so you can compare to exactly what's coded up.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:199]: I presume the tags inside the class mean they are private to the class...perhaps file an issue to replace char speeds with tensor char speeds everywhere, since they are computed a lot of places, mostly as `std::arrays`? I understand this would be outside the scope of this PR.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp:48]: I understand why you're multiplying by the char speeds while packaging instead of later, but then this makes the names of the variables confusing. I.e. `packaged_v_spacetime_metric` actually is `v_spacetime_metric * char_speeds[0]`. Maybe consider renaming the variables to things like `packaged_lambda_v_spacetime_metric`, or `packaged_char_speed_v_spacetime_metric` or something along those lines? Or at least a comment in the hpp file that these variables will actually not store what the name implies they will?

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp:153]: A comment clarifying how things change vs the equations in the help text because the normals point in opposite directions would be helpful, as would some comments regarding the signs. Since this empirically is stable, I presume it's all correct here, but I find it nontrivial to follow the signs here, to make sure they all are correct.

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Some suggestions for the docs, mostly. I looked at the hpp and cpp files, but did not yet look at the test.

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:109]: That's a great question! The mesh velocity is actually a correction to the shift, it's almost that (except this term!). The reason is that the equations are written as:
```
du_\alpha/dt + (B^i_{\alpha\beta} - v^i_g \delta_{\alpha\beta} ) d_i u_{\alpha}= S
```
The characteristic speeds are obtain by diagonalizing 
```
(B^i_{\alpha\beta} - v^i_g \delta_{\alpha\beta}) n_i
```
Since the second matrix is already diagonalization matrices (that take the evolved vars to the chars) are unchanged, but the char speeds are shifted by `-v^i_g n_i`. It just happens to work out so that pretty much all the shift vectors are adjusted by the velocity, except the one multiplying the `\gamma_1`. Does that clarify what's happening?

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:199]: I don't remember why exactly we use a `std::array` for the char speeds, but there was a lot of back and forth. Going forward I'm planning on having the boundary terms/numerical fluxes compute the char speeds internally so it should be less of a problem, but I'll file an issue to revisit once we know what's happening.

Update:
#2252

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp:153]: I'm adding the SpECTRE form into the dox instead

**Comment by nilsdeppe** [comment]
@geoffrey4444 thanks for the review! I've pushed a fixup

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:109]: Thanks! I think I get it now. You're subtracting a diagonal matrix from B^i_{\alpha\beta}, so that changes the eigenvalues but not the eigenvectors. Since it turns out that the char speeds except one all involve \beta^I n_i once, subtracting v^i_g n_i acts like changing the shift, but not quite, because of that one exception, the gamma_1 term.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.py
Comment (geoffrey4444) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.py]: Nitpick, but the help text says >=0, whereas here you use > 0. Since you're multiplying, the same thing happens for the case of 0, but maybe it would be nice to be consistent?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.py
Comment (geoffrey4444) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.py]: I thought we were calling this one v^g or v^spacetime_metric?

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.cpp:199]: Just a note: I checked that these (and the corresponding python functions) agree with what's in the help text

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (geoffrey4444) [inline src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:182]: Thanks! I presume the equations are correct, because the code works empirically, but I'm confused by the "+/- swap roles)...I would have expected this would give the opposite sign in the () terms for D_{\Phi} in the spectre version vs the "same direction normals" version?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Aside from double-checking the sign in D_Phi, just a couple of nitpick suggestions. I checked the test, and that also looks reasonable. I tried to figure out why my version (no PR, it's just a branch on my fork), which supposedly did the same thing, wasn't stable empirically, but I couldn't see the difference. Probably not worth hunting that down.

After you double-check the signs and consider my remaining nitpicks, I'm happy to approve

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (prayush) [reply src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:109]: Equivalently, the shift `N^i` is changed by `-v^i_g` everywhere it appears on its own contracting a `d_i u_\alpha`, except at this one place when `u_\alpha = \psi_{ab}` where it appears with a prefactor as `(1 + \gamma_1)N^k` and so the velocity gets absorbed by the first term

**Review by prayush — DISMISSED** [review]
This looks good to me, module things that @geoffrey4444 already noted above. Just one question: Why is this called the upwind *penalty* correction, i.e. are there penalty terms here in addition to the upwind corrections?

**Comment by nilsdeppe** [comment]
@prayush naming this was a struggle... There are no actual penalty terms (well, it's set to 1 because that's what DG tells us to do). The goal was to draw the analogy to the upwind multipenalty schemes, which this is but for DG. I'm not attached to the name so if you prefer something else I'm happy to change it. I also wanted to avoid upwind flux since it's not a flux but just a boundary correction term for the nonconservative products.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.py
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.py]: Sorry, this was a copy-paste error from the SW upwind scheme, changing to v^g

File: src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp:182]: This comes up because of the extra minus sign in the `\bar{\lambda}`, which is how it's written in the code and why I wrote it this way in the documentation

**Comment by nilsdeppe** [comment]
Okay, pushed another fixup changing the python code. Let me know if there's something I should clarify in the documentation

**Review by geoffrey4444 — COMMENTED** [review]
I’m happy for you to squash

**Comment by nilsdeppe** [comment]
Okay, rebased on develop and squashed. Thanks for the reviews :)

**Review by geoffrey4444 — DISMISSED** [review]

**Comment by kidder** [comment]
@prayush please look at

**Comment by nilsdeppe** [comment]
Rebased on develop and formatted the python file that this was complaining about

**Review by geoffrey4444 — DISMISSED** [review]

**Comment by geoffrey4444** [comment]
@prayush I think this is just waiting for your approval or further comments, so if you could take a look when you get a chance, that would be great!

**Review by prayush — DISMISSED** [review]

**Comment by kidder** [comment]
LGTM, rebase

**Comment by nilsdeppe** [comment]
Okay, rebased on develop. Thanks for the reviews :)

**Comment by geoffrey4444** [comment]
clang-tidy check failed

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@geoffrey4444 @prayush clang-tidy issues fixed

**Review by geoffrey4444 — APPROVED** [review]

**Review by prayush — APPROVED** [review]

**Pull request merged**
